File: coolmeal_app/lib/models/meal_plan.dart
class MealPlan {
  final int index;
  final String? breakfast;
  final String? lunch;
  final String? dinner;
  final String? combinedIngredients;
  final double? totalEnergy;
  final double? totalProtein;
  final double? totalFat;
  final double? totalCarbohydrates;
  final double? totalMagnesium;
  final double? totalSodium;
  final double? totalPotassium;
  final double? totalSaturatedFattyAcids;
  final double? totalMonounsaturatedFattyAcids;
  final double? totalPolyunsaturatedFattyAcids;
  final double? totalFreeSugar;
  final double? totalStarch;
  final double? price;

  MealPlan({
    required this.index,
    required this.breakfast,
    required this.lunch,
    required this.dinner,
    required this.combinedIngredients,
    required this.totalEnergy,
    required this.totalProtein,
    required this.totalFat,
    required this.totalCarbohydrates,
    required this.totalMagnesium,
    required this.totalSodium,
    required this.totalPotassium,
    required this.totalSaturatedFattyAcids,
    required this.totalMonounsaturatedFattyAcids,
    required this.totalPolyunsaturatedFattyAcids,
    required this.totalFreeSugar,
    required this.totalStarch,
    required this.price,
  });

  factory MealPlan.fromJson(Map<String?, dynamic> json) {
    return MealPlan(
      index: json['index'],
      breakfast: json['Breakfast'],
      lunch: json['Lunch'],
      dinner: json['Dinner'],
      combinedIngredients: json['Combined Ingredients'],
      totalEnergy: json['Total Energy(Kcal)'],
      totalProtein: json['Total Protein(g)'],
      totalFat: json['Total fat(g)'],
      totalCarbohydrates: json['Total Carbohydrates(g)'],
      totalMagnesium: json['Total Magnesium(mg)'],
      totalSodium: json['Total Sodium(mg)'],
      totalPotassium: json['Total Potassium(mg)'],
      totalSaturatedFattyAcids: json['Total Saturated Fatty Acids(mg)'],
      totalMonounsaturatedFattyAcids:
          json['Total Monounsaturated Fatty Acids(mg)'],
      totalPolyunsaturatedFattyAcids:
          json['Total Polyunsaturated Fatty Acids(mg)'],
      totalFreeSugar: json['Total Free Sugar(g)'],
      totalStarch: json['Total Starch(g)'],
      price: json['Price'],
    );
  }

  Map<String?, dynamic> toJson() {
    return {
      'index': index,
      'Breakfast': breakfast,
      'Lunch': lunch,
      'Dinner': dinner,
      'Combined Ingredients': combinedIngredients,
      'Total Energy(Kcal)': totalEnergy,
      'Total Protein(g)': totalProtein,
      'Total fat(g)': totalFat,
      'Total Carbohydrates(g)': totalCarbohydrates,
      'Total Magnesium(mg)': totalMagnesium,
      'Total Sodium(mg)': totalSodium,
      'Total Potassium(mg)': totalPotassium,
      'Total Saturated Fatty Acids(mg)': totalSaturatedFattyAcids,
      'Total Monounsaturated Fatty Acids(mg)': totalMonounsaturatedFattyAcids,
      'Total Polyunsaturated Fatty Acids(mg)': totalPolyunsaturatedFattyAcids,
      'Total Free Sugar(g)': totalFreeSugar,
      'Total Starch(g)': totalStarch,
      'Price': price,
    };
  }
}

File: server/schema/meal_plan.py
from pydantic import BaseModel, Field


# Pydantic model for data validation
class MealPlan(BaseModel):

    Breakfast: str = Field(..., description="Meal for breakfast")
    Lunch: str = Field(..., description="Meal for lunch")
    Dinner: str = Field(..., description="Meal for dinner")
    Price: float = Field(..., ge=0, description="Price of the meal plan")
    Total_Energy_Kcal: float = Field(
        ..., ge=0, alias="Total Energy(Kcal)", description="Total energy in kcal"
    )
    Total_Protein_g: float = Field(
        ..., ge=0, alias="Total Protein(g)", description="Total protein in grams"
    )
    Total_Fat_g: float = Field(
        ..., ge=0, alias="Total fat(g)", description="Total fat in grams"
    )
    Total_Carbohydrates_g: float = Field(
        ...,
        ge=0,
        alias="Total Carbohydrates(g)",
        description="Total carbohydrates in grams",
    )
    Total_Magnesium_mg: float = Field(
        ..., ge=0, alias="Total Magnesium(mg)", description="Total magnesium in mg"
    )
    Total_Sodium_mg: float = Field(
        ..., ge=0, alias="Total Sodium(mg)", description="Total sodium in mg"
    )
    Total_Potassium_mg: float = Field(
        ..., ge=0, alias="Total Potassium(mg)", description="Total potassium in mg"
    )
    Total_Saturated_Fatty_Acids_mg: float = Field(
        ...,
        ge=0,
        alias="Total Saturated Fatty Acids(mg)",
        description="Total saturated fatty acids in mg",
    )
    Total_Monounsaturated_Fatty_Acids_mg: float = Field(
        ...,
        ge=0,
        alias="Total Monounsaturated Fatty Acids(mg)",
        description="Total monounsaturated fatty acids in mg",
    )
    Total_Polyunsaturated_Fatty_Acids_mg: float = Field(
        ...,
        ge=0,
        alias="Total Polyunsaturated Fatty Acids(mg)",
        description="Total polyunsaturated fatty acids in mg",
    )
    Total_Free_Sugar_g: float = Field(
        ..., ge=0, alias="Total Free sugar(g)", description="Total free sugar in grams"
    )
    Total_Starch_g: float = Field(
        ..., ge=0, alias="Total Starch(g)", description="Total starch in grams"
    )
    index: int = Field(None, description="Index of the meal plan")

    class Config:
        allow_population_by_field_name = True
        schema_extra = {
            "example": {
                "Breakfast": "Red Rice with Fish Curry, Tapioca Chips, Batu Moju",
                "Lunch": "Red Rice with Banana Blossom Curry, Green Gram Curry, Fenugreek Curry",
                "Dinner": "Vegetable Kottu with Devilled Fish",
                "Price": 792.3,
                "Total Energy(Kcal)": 1482.37,
                "Total Protein(g)": 73.5357,
                "Total fat(g)": 83.1566,
                "Total Carbohydrates(g)": 175.9371,
                "Total Magnesium(mg)": 478.5,
                "Total Sodium(mg)": 128.367,
                "Total Potassium(mg)": 3346.98,
                "Total Saturated Fatty Acids(mg)": 37105.9815,
                "Total Monounsaturated Fatty Acids(mg)": 6314.2741,
                "Total Polyunsaturated Fatty Acids(mg)": 5152.7615,
                "Total Free sugar(g)": 11.7442,
                "Total Starch(g)": 151.1037,
                "index": 2,
            }
        }
